### ISSUE TITLE
Support other FIWARE brokers

### ISSUE THREAD

**Issue by dkastl** [gtt-project/redmine_gtt_fiware#16]
There are other FIWARE brokers, which may have limited API aupport

- [ ] Orion Context Broker:
- [ ] Orion-LD Context Broker
- [ ] Scorpio Broker
- [ ] Stellio Context Broker

**Comment by dkastl** [comment]
The required database fields should be available already.

**Comment by dkastl** [comment]
Status after #86: NGSI-LD subscription support is implemented (per-standard `SubscriptionRequest` builder; `/ngsi-ld/v1/` endpoints, `notificationTrigger`, `@context`, `endpoint`+`receiverInfo`), and the notification pipeline handles both NGSI-LD and NGSIv2 entities through one code path. What keeps this open: live verification against GeonicDB (reference broker) and Orion-LD portability — unit/functional coverage exists but no real broker round-trip has been run yet. The v3.0 tag waits on this.

**Comment by dkastl** [comment]
## Live verification test plan (GeonicDB + Orion-LD)

This is the v3.0 gate: everything below is covered by unit/functional tests against stubs; this plan verifies the same behaviour against real brokers. Reference broker: **GeonicDB** (hosted, NGSI-LD native). Portability check: **Orion-LD** (local container).

### 0. Environment / preflight

- [ ] Local Redmine (podman stack) with `redmine_gtt` + `redmine_gtt_fiware` at current `main`, migrations applied.
- [ ] **Public callback URL**: notifications must reach Redmine, and the callback URL is built from `request.base_url`, so publishing must happen through a publicly reachable hostname. Use a Cloudflare quick tunnel (`cloudflared tunnel --url http://localhost:3000`) and do all publishing through the tunnel URL. One tunnel serves both brokers.
- [ ] GeonicDB token available (`GEONICDB_TOKEN`); create a **stored-auth** broker connection in Administration → FIWARE Broker Connections (`standard: NGSI-LD`, URL `https://api.geonicdb.geolonia.com`, context = core context or tenant context).
- [ ] Orion-LD + MongoDB running locally (`quay.io/fiware/orion-ld`), second broker connection (no auth or stored token).
- [ ] A test project with the `gtt` and `gtt_fiware` modules enabled and a member with `add_issues` + `manage_subscription_templates`.

### 1. Subscription lifecycle (per broker)

| # | Step | Expected |
| --- | --- | --- |
| 1.1 | Publish an NGSI-LD template (stored auth) | Server-side POST to `/ngsi-ld/v1/subscriptions` with `Content-Type: application/ld+json`; broker returns 201; `subscription_id` stored from the `Location` header; token never visible in the browser |
| 1.2 | `GET /ngsi-ld/v1/subscriptions/<id>` via curl | Payload matches the builder output: `entities`, `notificationTrigger`, `@context`, `notification.endpoint.uri` = tunnel callback, `receiverInfo` carries `X-Gtt-Webhook-Secret` + registration URL |
| 1.3 | Sync button with the subscription live | Status reconciled (`active`), no state change |
| 1.4 | Delete the subscription broker-side via curl, then Sync | Local `subscription_id` cleared, template shows as unpublished |
| 1.5 | Re-publish, then Unpublish from Redmine | Broker returns 404 on subsequent GET; local id cleared |
| 1.6 | Browser-mode connection publish (Orion-LD, no auth) | JS flow works end to end; same payload accepted |

### 2. Notification → issue pipeline (the critical path)

| # | Step | Expected |
| --- | --- | --- |
| 2.1 | Create a matching entity (e.g. `TemperatureSensor` with `temperature` Property + `location` GeoProperty) | **`receiverInfo` is forwarded as HTTP headers** — this is the make-or-break item; brokers vary. Issue created; subject/description rendered from `${id}` / `${attrs.temperature.value}`; `geom` set from the GeoProperty; author = template member |
| 2.2 | PATCH the entity's `temperature` within the `threshold_create` window | No new issue; journal note (or blank-notes journal) on the existing issue |
| 2.3 | PATCH the entity's `location` | `geom` updated + `geom` journal detail |
| 2.4 | Wait past the threshold (set it to 0–1h for testing), PATCH again | New issue created |
| 2.5 | curl the notification endpoint directly with a wrong secret | 401, no issue |
| 2.6 | Malformed JSON body via curl | 400 |

### 3. Subscription filters

| # | Step | Expected |
| --- | --- | --- |
| 3.1 | `watchedAttributes` (template `attrs`) set to `temperature`; PATCH a different attribute | No notification |
| 3.2 | `q` filter `temperature>30`; create entities at 25 and 35 | Only the 35 notifies |
| 3.3 | `geoQ` near/Point and within/Polygon (project boundary insert) | Only entities inside notify; confirms the lowercase→GeoJSON geometry-name mapping works against a real broker |
| 3.4 | `notificationTrigger` = `entityCreated` only; PATCH an existing entity | No notification on update |

### 4. Tenancy & context (GeonicDB specifics)

- [ ] If the GeonicDB tenant model uses `NGSILD-Tenant`: set `fiware_service` on the connection and confirm subscription + notifications stay inside the tenant.
- [ ] Custom `@context` (tenant vocabulary) on the connection; template override on one template; confirm term expansion works and notifications still parse (the Entity normalizer treats keys literally — compacted vs expanded attribute names in notifications is a known thing to observe).

### 5. Known open questions to answer during the run

1. **Does the broker forward `receiverInfo` as headers?** (Orion-LD: yes since 1.x; GeonicDB: to confirm.) If not, webhook auth needs a fallback (e.g. secret in the callback URL path/query) — that would be a follow-up issue.
2. **Notification format**: we subscribe with `format: normalized`; confirm entities arrive as Property/Relationship/GeoProperty maps (what `Entity#from_ngsi_ld` expects), not keyValues.
3. **`oneshot`**: NGSIv2-only concept; an LD template with status `oneshot` currently publishes `isActive: false`. Decide whether to hide `oneshot` for LD templates (form work, #66) or map it differently.
4. **`expiresAt`** roundtrip and what Sync reports after expiry (`expired` → local `inactive`).

### Result recording

One comment per broker on this issue with the table filled in (pass/fail/notes), plus follow-up issues for anything that fails. When both brokers pass §1–§2 (lifecycle + notification pipeline), this issue closes and **v3.0 can be tagged**.

**Comment by dkastl** [comment]
## GeonicDB live verification: results (2026-07-28)

Run against the dedicated instance (`nznqqrklzl.geonicdb.jp`, scratch tenant `gtt_fiware_dev`), notifications delivered to a local listener through a cloudflared tunnel. Plugin code exercised directly (the real `SubscriptionRequest`, `Entity` and `TemplateRenderer` classes), not reimplemented.

### Passed

| Area | Result |
| --- | --- |
| Subscription create | Our builder's payload accepted as-is, `201`, at `/ngsi-ld/v1/subscriptions` with `Content-Type: application/ld+json` and `NGSILD-Tenant` |
| **`receiverInfo` -> HTTP headers** | **Confirmed.** `X-Gtt-Webhook-Secret` arrived as a real request header on the notification. The whole webhook-auth design (#58) works against this broker |
| Notification envelope | `{subscriptionId, data:[...], notifiedAt, triggerReason}` - our controller's `data[]` parsing is correct |
| Entity normalization | `Entity` + `TemplateRenderer` render correctly from the **live notification payload**: subject/description/notes, `${id}`, `${attrs.x.value}`, shorthand, missing -> empty |
| GeoProperty -> geom | `${location}` resolves structurally to valid GeoJSON `Point` |
| `watchedAttributes` | Enforced: patching a non-watched attribute produced no notification |
| Tenant naming | `gtt_fiware_dev` / `gtt_fiware_demo` pass the plugin's `#37` validation unchanged (underscores allowed) |
| Entity geoquery | `georel=near;maxDistance` filters correctly on **queries** (1 match near, 0 far) |

### Broker-side limitations found (not plugin bugs)

**`q` and `geoQ` are accepted on a subscription but not applied to notifications.**

- A subscription whose `geoQ` points at Osaka with `maxDistance==1000` still fired for an entity in Tokyo (~400 km away). The **identical** geoQ as an entity *query* correctly returns 0 matches, so the geo engine works; it just is not applied to notification matching.
- A subscription with `q: "severity>100"` (never true) fired on every update.
- Both are accepted with `201`, so this fails silently: a user who configures an area or a query filter receives everything.

Consequence for us: the Filters section of the form (#66) offers `watchedAttributes`, `q` and geo area. Only `watchedAttributes` currently does anything on GeonicDB. Options: report upstream to the GeonicDB team, and/or post-filter plugin-side in `NotificationProcessor` (we already have the entity geometry and attribute values). I have not implemented either - worth a decision first.

### Notes / gotchas

- The **notification** payload types differ from the entity `GET`: attributes come back as `String` / `Number` rather than `Property`, and a `Relationship` carries **`value`, not `object`**. Our normalizer handles it because it falls back to `value` when `object` is absent, but it is worth a regression test with the real shape.
- `geoQ.coordinates` accepts both a JSON array and a string at creation. The NGSIv2-style `"lat, lon"` string is rejected on queries (`400 Latitude out of range`) but silently accepted on subscriptions. Since geoQ is ignored anyway this is currently moot, but the form's coords placeholder is NGSIv2-shaped (`lat1, lon1; ...`) while NGSI-LD wants GeoJSON `[lon, lat]` order - a real trap once geoQ works.
- **NGSIv2 cannot be verified here**: the API key is scoped to `/ngsi-ld/**` and `/v2/*` returns 403 by design, and entities are protocol-isolated. Verifying #65 needs a local Orion (v2) or a widened policy.

### Status

Sections 1, 2 and the entity/tenant parts of 4 of the plan pass. Remaining before this issue closes: Orion-LD portability, and a decision on the `q`/`geoQ` gap. Scratch tenant left clean (0 subscriptions); the demo tenant was not touched.

**Comment by dkastl** [comment]
### Follow-up: the `q` / `geoQ` gap is a broker bug, confirmed against the reference implementation

Filed upstream as **geolonia/geonicdb#1570**.

Before filing I ruled out our own code as the cause: the reproduction is **pure curl with hand-written minimal JSON**, no plugin involvement, and the identical script was run against **Orion-LD 1.12.0** (FIWARE reference implementation, local container):

| Subscription | Orion-LD 1.12.0 | GeonicDB |
| --- | --- | --- |
| no filter (control) | notified | notified |
| `geoQ` 400 km away | **suppressed** | **notified** |
| `q: severity>100` | **suppressed** | **notified** |

The control firing on both proves the test rig is sound. Spec side is corroborated by the ScorpioBroker docs (which document `geoQ` in subscriptions with a near-identical `near;maxDistance==2000` example) and the FIWARE NGSI-LD tutorials for `q`, plus GeonicDB's own ETSI GS CIM 009 V1.9.1 compliance epic.

So: **nothing to fix on the plugin side** for this. Our builder emits a correct, spec-shaped `geoQ`/`q` that the reference implementation honours. Once the broker fix lands, our filters start working with no change from us.

Open question for us to decide separately: whether to post-filter plugin-side in `NotificationProcessor` as a defensive measure (we already hold the entity geometry and attribute values), or to rely on the broker. I would lean towards waiting for the upstream fix rather than duplicating broker semantics in the plugin.

This also completes the Orion-LD portability half of this issue: our subscription payload was accepted (`201`) and behaved correctly there too.

**Comment by dkastl** [comment]
## NGSIv2 live verification — local Orion v2 (2026-07-28)

Completes the NGSIv2 half (the GeonicDB key is NGSI-LD-scoped, so this ran against a local broker).

**Setup:** `fiware/orion:3.12.0` + `mongo:5.0` (podman, same network as the dev Redmine container), tenant `roadinspection`, open broker (stored connection with no token — also exercises the token-less stored path). Template created through the UI against the v2 connection.

### Results — all PASSED

| Check | Result |
| --- | --- |
| v2 form toggles (#66) | alteration labels show v2 names (`entityCreate`/`entityChange`/…), `notify_on_metadata_change` visible, `@context` hidden |
| Subscription create | `POST /v2/subscriptions` → **201**, builder payload accepted as-is |
| Payload shape | `subject.condition.expression.q`, `alterationTypes`, `notifyOnMetadataChange`, `throttling`, `httpCustom` with `X-Gtt-Webhook-Secret` + registration URL header |
| httpCustom headers forwarded | notification authenticated by the webhook secret → issue created (headers demonstrably arrive) |
| Notification envelope | `{subscriptionId, data: [...]}` handled by the controller |
| v2 normalized attrs | `Text`/`Number`/`geo:json` render through Entity + TemplateRenderer (subject/description correct) |
| `geo:json` location | issue geometry set, point on the map |
| **`q` filter enforcement** | three-phase test: create severity=2 → suppressed; update to 3 → suppressed; update to 5 → **exactly one** notification (`timesSent: 1`, `lastSuccessCode: 200`) → issue created |
| Sync (#13) | `GET /v2/subscriptions/:id` status `active` recognized, local status consistent |
| Unpublish | `DELETE` → broker subscription list empty, local `subscription_id` cleared |

### Notable

- **Orion v2 applies `q` to notifications correctly** — the same test that GeonicDB fails silently (geolonia/geonicdb#1570). This confirms the gap is broker-side, not in our subscription payloads.
- **Gotcha:** Orion rejects `httpCustom.url` hosts containing underscores (`400 invalid custom /url/`) — hit it with the container name `srv_redmine_redmine_1`, worked via the underscore-free alias. One more reason for #101 (explicitly configured callback base URL).
- Not covered here: `oneshot` status round-trip, v2 attachment download, long-run throttling behaviour.

With the NGSI-LD half verified earlier against GeonicDB, both standards have now passed a full live round-trip (template → publish → broker notification → authenticated issue creation with geometry).

**Comment by dkastl** [comment]
Closing as complete — both NGSI standards have passed a full live round-trip, and the procedure is now permanent:

- **NGSI-LD** (GeonicDB, 2026-07-28): subscription create, `receiverInfo` forwarded as HTTP headers (webhook auth works), notification envelope, entity/template rendering incl. `${location}` geometry, `watchedAttributes` enforcement — plus one upstream broker gap found and reported (geolonia/geonicdb#1570: `q`/`geoQ` accepted but not applied to notifications).
- **NGSIv2** (local Orion 3.12.0, 2026-07-28/29): see the [results table above](https://github.com/gtt-project/redmine_gtt_fiware/issues/16#issuecomment-5102056738) — all passed, including the three-phase `q` filter test, sync and unpublish.
- **All three auth modes** (stored / relayed / browser) verified end-to-end through the real UI against a header-echo broker on 2026-07-29, which found and fixed #107.

The whole rig is checked in since #108 as [`doc/release_verification.md`](https://github.com/gtt-project/redmine_gtt_fiware/blob/main/doc/release_verification.md) + `doc/scripts/`, so this verification is repeatable for every future release. Remaining nice-to-haves (oneshot round-trip, v2 attachment download, long-run throttling) are noted there and don't block v3.0.
